Plugin crash when trying to pass a computed value for a mac address

When trying to set a vm's nic's mac address, the xenorchestra provider does not handle "computed" values because of the validation done [on the mac address field](https://github.com/terra-farm/terraform-provider-xenorchestra/blob/51c49b950466c47befa6860263ac8a47e9da55a7/xoa/resource_xenorchestra_vm.go#L237), meaning that the mac's value has to be known ahead of time.

Terraform example code where the value is computed:

```hcl
# provider ivoronin/macaddress:v0.3.0
resource "macaddress" "mac" {
  prefix = [0, 22, 62] # => example value for the resulting address: "00:16:3e:b6:8b:8b"
}

resource "xenorchestra_vm" "vm" {
  [...]
  network {
    network_id  = data.xenorchestra_network.lookup.id
    mac_address = macaddress.mac.address
  }
 [...]
}
```

```bash
terraform plan
╷
│ Error: Plugin did not respond
│
│   with module.vm.xenorchestra_vm.vm,
│   on ../main.tf line 36, in resource "xenorchestra_vm" "vm":
│   36: resource "xenorchestra_vm" "vm" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).PlanResourceChange call. The plugin logs may   
│ contain more details.
╵

Stack trace from the terraform-provider-xenorchestra_v0.21.0 plugin:

panic: Mac address `74D93920-ED26-11E3-AC10-0800200C9A66` was not parsable. This should never happened because Terraform's validation should happen before this is stored into state

goroutine 179 [running]:
github.com/ddelnano/terraform-provider-xenorchestra/xoa.resourceVmSchema.func1(0xd752e0, 0xc000349de0, 0x15, 0x0)
        github.com/ddelnano/terraform-provider-xenorchestra/xoa/resource_xenorchestra_vm.go:185 +0x145
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diffString(0xc000402810, 0xc000478360, 0x15, 0xc0000e5e00, 0xc00034cf00, 0x1047b68, 0xc00077a380, 0x476d00, 0x7, 0x100000000000000)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/schema.go:1375 +0x6e3
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diff(0xc000402810, 0xc000478360, 0x15, 0xc0000e5e00, 0xc00034d508, 0x1047b68, 0xc00077a380, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/schema.go:957 +0x4e8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diffList(0xc000402810, 0xecb8c0, 0x7, 0xc00029f680, 0xc00034d508, 0x1047b68, 0xc00077a380, 0x203000, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/schema.go:1091 +0x631
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diff(0xc000402810, 0xecb8c0, 0x7, 0xc00029f680, 0xc000401960, 0x1047b68, 0xc00077a380, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/schema.go:959 +0x473
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.Diff(0xc000402810, 0x1043688, 0xc000623580, 0xc00033acb0, 0xc000854420, 0x0, 0xec45a0, 0xc0001dc1b0, 0x156b300, 0xd752e0, ...)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/schema.go:522 +0x219
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).SimpleDiff(0xc0001849c0, 0x1043688, 0xc000623580, 0xc00033acb0, 0xc000854420, 0xec45a0, 0xc0001dc1b0, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/resource.go:446 +0x9f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).PlanResourceChange(0xc00000c468, 0x1043688, 0xc000623580, 0xc000857c20, 0xc000623580, 0xe85aa0, 0xc00090af00)
        google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
Error: The terraform-provider-xenorchestra_v0.21.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

If XO will not accept invalid values for mac addresses, I don't think we need to have a custom validation here. We should simply forward the XO api error response to the user when they try to pass erroneous data.

--- a/go.mod
+++ b/go.mod
@@ -13,12 +13,13 @@ require (
 	github.com/Masterminds/semver/v3 v3.1.1 // indirect
 	github.com/Masterminds/sprig/v3 v3.2.2 // indirect
 	github.com/ProtonMail/go-crypto v0.0.0-20230828082145-3c4c8a2d2371 // indirect
-	github.com/agext/levenshtein v1.2.2 // indirect
+	github.com/agext/levenshtein v1.2.3 // indirect
 	github.com/apparentlymart/go-textseg/v15 v15.0.0 // indirect
 	github.com/armon/go-radix v1.0.0 // indirect
 	github.com/bgentry/speakeasy v0.1.0 // indirect
 	github.com/cenkalti/backoff/v3 v3.2.2 // indirect
 	github.com/cloudflare/circl v1.3.7 // indirect
+	github.com/ddelnano/terraform-provider-macaddress v0.0.0-20240110221909-992cb1a7c000 // indirect
 	github.com/fatih/color v1.13.0 // indirect
 	github.com/golang/protobuf v1.5.3 // indirect
 	github.com/google/go-cmp v0.6.0 // indirect
@@ -50,10 +51,10 @@ require (
 	github.com/mitchellh/cli v1.1.5 // indirect
 	github.com/mitchellh/copystructure v1.2.0 // indirect
 	github.com/mitchellh/go-testing-interface v1.14.1 // indirect
-	github.com/mitchellh/go-wordwrap v1.0.0 // indirect
+	github.com/mitchellh/go-wordwrap v1.0.1 // indirect
 	github.com/mitchellh/mapstructure v1.5.0 // indirect
 	github.com/mitchellh/reflectwalk v1.0.2 // indirect
-	github.com/oklog/run v1.0.0 // indirect
+	github.com/oklog/run v1.1.0 // indirect
 	github.com/posener/complete v1.2.3 // indirect
 	github.com/russross/blackfriday v1.6.0 // indirect
 	github.com/shopspring/decimal v1.3.1 // indirect
@@ -73,6 +74,9 @@ require (
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20231002182017-d307bd883b97 // indirect
 	google.golang.org/grpc v1.60.0 // indirect
 	google.golang.org/protobuf v1.31.0 // indirect
+	terraform-provider-macaddress v0.1.0 // indirect
 )
 
 replace github.com/ddelnano/terraform-provider-xenorchestra/client => ./client
+
+replace terraform-provider-macaddress v0.1.0 => github.com/ddelnano/terraform-provider-macaddress v0.1.0

--- a/xoa/provider_test.go
+++ b/xoa/provider_test.go
@@ -2,6 +2,7 @@ package xoa
 
 import (
 	"os"
+	"terraform-provider-macaddress/macaddress"
 	"testing"
 
 	"github.com/ddelnano/terraform-provider-xenorchestra/xoa/internal"
@@ -20,6 +21,7 @@ func init() {
 	testAccProvider = Provider()
 	testAccProviders = map[string]*schema.Provider{
 		"xenorchestra": testAccProvider,
+		"macaddress":   macaddress.Provider(),
 	}
 
 	testAccFailToStartHaltVmProvider = Provider()

--- a/xoa/resource_xenorchestra_vm.go
+++ b/xoa/resource_xenorchestra_vm.go
@@ -316,20 +316,11 @@ $ xo-cli xo.getAllObjects filter='json:{"id": "cf7b5d7d-3cd5-6b7c-5025-5c935c8cd
 						Required:    true,
 					},
 					"mac_address": &schema.Schema{
-						Type:        schema.TypeString,
-						Optional:    true,
-						Computed:    true,
-						Description: "The mac address of the network interface. This must be parsable by go's [net.ParseMAC function](https://golang.org/pkg/net/#ParseMAC). All mac addresses are stored in Terraform's state with [HardwareAddr's string representation](https://golang.org/pkg/net/#HardwareAddr.String) i.e. 00:00:5e:00:53:01",
-						StateFunc: func(val interface{}) string {
-
-							unformattedMac := val.(string)
-							mac, err := net.ParseMAC(unformattedMac)
-							if err != nil {
-								panic(fmt.Sprintf("Mac address `%s` was not parsable. This should never happened because Terraform's validation should happen before this is stored into state", unformattedMac))
-							}
-							return mac.String()
-
-						},
+						Type:             schema.TypeString,
+						Optional:         true,
+						Computed:         true,
+						Description:      "The mac address of the network interface. This must be parsable by go's [net.ParseMAC function](https://golang.org/pkg/net/#ParseMAC). All mac addresses are stored in Terraform's state with [HardwareAddr's string representation](https://golang.org/pkg/net/#HardwareAddr.String) i.e. 00:00:5e:00:53:01",
+						DiffSuppressFunc: suppressEquivalentMAC,
 						ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
 							mac_address := val.(string)
 							if _, err := net.ParseMAC(mac_address); err != nil {
@@ -1412,6 +1403,22 @@ func extractIpsFromNetworks(networks map[string]string) []guestNetwork {
 	}
 	log.Printf("[DEBUG] Extracted the following network interface ips: %v\n", devices)
 	return devices
+}
+
+func suppressEquivalentMAC(k, old, new string, d *schema.ResourceData) (suppress bool) {
+	fmt.Printf("[DEBUG] Comparing MACs old=%s new=%s\n", old, new)
+	if old == "" {
+		return false
+	}
+	newMAC, err := net.ParseMAC(new)
+	if err != nil {
+		return false
+	}
+	oldMAC, err := net.ParseMAC(old)
+	if err != nil {
+		return false
+	}
+	return oldMAC.String() == newMAC.String()
 }
 
 func suppressAttachedDiffWhenHalted(k, old, new string, d *schema.ResourceData) (suppress bool) {

--- a/xoa/resource_xenorchestra_vm_test.go
+++ b/xoa/resource_xenorchestra_vm_test.go
@@ -843,6 +843,47 @@ func TestAccXenorchestraVm_createWithCloudInitNetworkConfig(t *testing.T) {
 	})
 }
 
+func TestAccXenorchestraVm_createWithInvalidMacAddress(t *testing.T) {
+	vmName := fmt.Sprintf("%s - %s", accTestPrefix, t.Name())
+	invalidMac := "00-0a:83-b1:c0-01"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckXenorchestraVmDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccVmConfigWithMacAddress(vmName, invalidMac),
+				ExpectError: regexp.MustCompile(`Mac Address is invalid`),
+			},
+		},
+	})
+}
+
+func TestAccXenorchestraVm_createWithSentinelPlanValue(t *testing.T) {
+	resourceName := "xenorchestra_vm.bar"
+	vmName := fmt.Sprintf("%s - %s", accTestPrefix, t.Name())
+	// sentinelMac := "74D93920-ED26-11E3-AC10-0800200C9A66"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckXenorchestraVmDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccVmConfigWithMacAddressSentinelInput(vmName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccVmExists(resourceName),
+					resource.TestCheckResourceAttrSet(resourceName, "id"),
+					resource.TestCheckResourceAttr(resourceName, "network.#", "1"),
+					// internal.TestCheckTypeSetElemNestedAttrs(resourceName, "network.*", map[string]string{
+					// 	// All mac addresses should be formatted to use colons
+					// 	"mac_address": getFormattedMac(macWithDashes),
+					// }),
+					internal.TestCheckTypeSetElemAttrPair(resourceName, "network.*.*", "data.xenorchestra_network.network", "id")),
+			},
+		},
+	})
+}
+
 func TestAccXenorchestraVm_createWithDashedMacAddress(t *testing.T) {
 	resourceName := "xenorchestra_vm.bar"
 	vmName := fmt.Sprintf("%s - %s", accTestPrefix, t.Name())
@@ -2342,6 +2383,36 @@ resource "xenorchestra_vm" "bar" {
     }
 }
 `, accDefaultNetwork.NameLabel, accTestPool.Id, vmName, macAddress, accDefaultSr.Id)
+}
+
+func testAccVmConfigWithMacAddressSentinelInput(vmName string) string {
+	return testAccCloudConfigConfig(fmt.Sprintf("vm-template-%s", vmName), "template") + testAccTemplateConfig() + fmt.Sprintf(`
+resource "macaddress" "mac" {}
+
+data "xenorchestra_network" "network" {
+    name_label = "%s"
+    pool_id = "%s"
+}
+
+resource "xenorchestra_vm" "bar" {
+    memory_max = 4295000000
+    cpus  = 1
+    cloud_config = "${xenorchestra_cloud_config.bar.template}"
+    name_label = "%s"
+    name_description = "description"
+    template = "${data.xenorchestra_template.template.id}"
+    network {
+	network_id = "${data.xenorchestra_network.network.id}"
+	mac_address = macaddress.mac.address
+    }
+
+    disk {
+      sr_id = "%s"
+      name_label = "disk 1"
+      size = 10001317888
+    }
+}
+`, accDefaultNetwork.NameLabel, accTestPool.Id, vmName, accDefaultSr.Id)
 }
 
 func testAccVmConfigWithTwoMacAddresses(vmName, firstMac, secondMac string) string {

--- a/xoa/resource_xenorchestra_vm_test.go
+++ b/xoa/resource_xenorchestra_vm_test.go
@@ -862,7 +862,6 @@ func TestAccXenorchestraVm_createWithInvalidMacAddress(t *testing.T) {
 func TestAccXenorchestraVm_createWithSentinelPlanValue(t *testing.T) {
 	resourceName := "xenorchestra_vm.bar"
 	vmName := fmt.Sprintf("%s - %s", accTestPrefix, t.Name())
-	// sentinelMac := "74D93920-ED26-11E3-AC10-0800200C9A66"
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
 		Providers:    testAccProviders,
@@ -874,10 +873,6 @@ func TestAccXenorchestraVm_createWithSentinelPlanValue(t *testing.T) {
 					testAccVmExists(resourceName),
 					resource.TestCheckResourceAttrSet(resourceName, "id"),
 					resource.TestCheckResourceAttr(resourceName, "network.#", "1"),
-					// internal.TestCheckTypeSetElemNestedAttrs(resourceName, "network.*", map[string]string{
-					// 	// All mac addresses should be formatted to use colons
-					// 	"mac_address": getFormattedMac(macWithDashes),
-					// }),
 					internal.TestCheckTypeSetElemAttrPair(resourceName, "network.*.*", "data.xenorchestra_network.network", "id")),
 			},
 		},

